Ogurtsov/gcs proxy server dev

OgurtsovAndrei opened the pull request.

OgurtcovAndrey committed:
Impl io.Reader on SGB

OgurtcovAndrey committed:
Add responses for requests

OgurtcovAndrey committed:
Create APIs:

- to Build & Read request
- to send messages in Connection Group
- to receive messages in Server Connection Group

OgurtcovAndrey committed:
Create 1 conn UnreliableGcsWriter & Proxy server

OgurtcovAndrey committed:
Create N conn UnreliableGcsWriter & Proxy server

dominiquelefevre commented on proxy/GSCProxyCommon.go:
```diff
@@ -44,3 +187,145 @@ func RequestTypeToString(requestType uint32) string {
 	}
 }
+
+func readRequestHeader(reader io.Reader) (RequestHeader, error) {
+	var header RequestHeader
+	err := binary.Read(reader, binary.BigEndian, &header)
+	if err != nil {
+		return header, err
+	}
+	return header, nil
```
A simpler and more idiomatic way is
```
func readRequestHeader(r io.Reader) (hdr RequestHeader, err error) {
    err = binary.Read(r, binary.BigEndian, &hdr)
    return hdr, err
}
```

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		log.Printf("RequestMessage: Error writing header: %v", err)
+		return 0, err
+	}
```
What is the point of this code?

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+		return 0, err
+	}
+	copy(p, headerBuf)
+	n += len(headerBuf)
```
Something is very broken here. You've only allocated `headerBuf`, but never copied anything into it, so `copy(p, headerBuf)` only zeroes `len(headerBuf)` first bytes.

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+type ConnectionGroup struct {
+	messages       chan *RequestMessage
+	responseMap    map[uint32]chan *ResponseMessage
+	responseMapMux sync.Mutex
```
"mux" typically stands for "muxer" like `http.ServeMux".

`responseMaxLock`

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+
+type ConnectionGroup struct {
+	messages       chan *RequestMessage
+	responseMap    map[uint32]chan *ResponseMessage
```
Such definition of `responseMap` is not going to evolve nicely. For now it suffices to have `chan *ResponseMessage` instead of `struct PendingResponse` or `struct Request`, but you are going to need a struct when you decide to implement request cancelation.

I think `map[uint32]*Request` (the name `struct Request` may be different) is preferable as you will be able to create an API similar to that of `http.Client`:
```
req := gcsClient.NewRequest(...)
resp := gcsClient.Do(req)
...
res, err := resp.Wait()
...
```

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+		nConnections: nConnections,
+	}
+
+	for i := 0; i < nConnections; i++ {
+		go func(i int) {
+			log.Printf("Starting connection goroutine %d", i)
+			if err := cg.handleConnection(); err != nil {
+				log.Printf("Connection goroutine %d error: %v", i, err)
+			}
+		}(i)
+	}
```
1. There is no way to implement `ConnectionGroup.Close()` because you cannot track the termination of all per-connection goroutines. As we've discussed earlier, goroutines need to be part of a workgroup.
2. This code has no error handling. If any connection attempt fails, you just get a connpool with fewer connections. What happens when all connections die?

dominiquelefevre commented on writers/ScatterGatherBuffer.go:
```diff
@@ -177,3 +177,44 @@ func (sgb *ScatterGatherBuffer) DropFirst(amount uint32) {
 	sgb.size -= dropped
 }
+
+func (sgb *ScatterGatherBuffer) TakeFirstUnsafe(n uint32) *ScatterGatherBuffer {
```
Don't use words that are too generic. What does "unsafe" even mean in this context?

dominiquelefevre commented on proxy/GcsConnection.go:
```diff
@@ -0,0 +1,300 @@
+			log.Printf("writeToConn: Sending request for RequestUid=%d", req.header.RequestUid)
+			if _, err := io.Copy(conn, req); err != nil {
+				log.Printf("writeToConn: Error writing request: %v", err)
+				writeErrCh <- err
```
This notifies only `handleConnection()` about an error. How do you fail the `req`?

OgurtcovAndrey committed:
Minor fixes

OgurtcovAndrey committed:
Add Err handing

OgurtcovAndrey committed:
Fixup retries in ReliableWriter.go

OgurtcovAndrey committed:
Create new connection pool on server side

OgurtsovAndrei commented:
@dominiquelefevre 
I created a new error handling mechanism
And create some proto, where I extract connection pool for 1 client from the global server connection group
It's a proto, so I still have to debug

OgurtcovAndrey committed:
Fix handshake bug

- Now we SEND handshake OK response to connection, not to connPool
(Before he had handshake response delivery mismatches)
